mineral_deposit_tracking/src/mineral_deposit_tracker.cpp

```diff
@@ -0,0 +1,124 @@
+// THE SOFTWARE.
+
+#include <tf2_ros/transform_listener.h>
+#include <tf2_eigen/tf2_eigen.h>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+#include <stsl_interfaces/msg/mineral_deposit_array.hpp>
+#include <stsl_interfaces/srv/reset_mineral_deposit_tracking.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include "kalman_filter.hpp"
+
+namespace mineral_deposit_tracking
+{
+
+class MineralDepositTracker : public rclcpp::Node
+{
+public:
+  explicit MineralDepositTracker(const rclcpp::NodeOptions & options)
+  : rclcpp::Node("mineral_deposit_tracker", options),
+    tf_buffer_(get_clock()),
+    tf_listener_(tf_buffer_),
+    filter_(Eigen::Matrix2d::Identity(),
+      Eigen::Matrix2d::Identity() * 1e-4, Eigen::Matrix2d::Identity())
+  {
+    tracked_deposit_publisher_ = create_publisher<geometry_msgs::msg::PoseStamped>(
+      "~/tracked_deposit", rclcpp::SystemDefaultsQoS());
+    deposit_subscription_ = create_subscription<stsl_interfaces::msg::MineralDepositArray>(
+      "~/sensed_deposits", rclcpp::SystemDefaultsQoS(),
+      std::bind(&MineralDepositTracker::DepositMeasurementCallback, this, std::placeholders::_1));
+    reset_service_ = create_service<stsl_interfaces::srv::ResetMineralDepositTracking>(
+      "~/reset", std::bind(
+        &MineralDepositTracker::ResetCallback, this, std::placeholders::_1,
+        std::
+        placeholders::_2));
+  }
+
+private:
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr tracked_deposit_publisher_;
+  rclcpp::Subscription<stsl_interfaces::msg::MineralDepositArray>::SharedPtr deposit_subscription_;
+  rclcpp::Service<stsl_interfaces::srv::ResetMineralDepositTracking>::SharedPtr reset_service_;
+  int deposit_id_;
+  KalmanFilter filter_;
+
+  void DepositMeasurementCallback(const stsl_interfaces::msg::MineralDepositArray::SharedPtr msg)
+  {
+    try {
+      filter_.TimeUpdate();
+      const auto found_deposit = std::find_if(
+        msg->deposits.begin(), msg->deposits.end(), [this](const auto & deposit) {
+          return deposit.id == deposit_id_;
+        });
+      if (found_deposit == msg->deposits.end()) {
+        return;
+      }
+      const auto & detection = *found_deposit;
+
+      const Eigen::Vector3d sensor_frame_position = detection.range * Eigen::Vector3d{std::cos(
+          detection.heading), std::sin(detection.heading), 0.0};
+
+      const auto transform =
+        tf_buffer_.lookupTransform(
+        "map", msg->header.frame_id, msg->header.stamp, rclcpp::Duration::from_seconds(
+          0.1));
+
+      Eigen::Vector3d map_frame_position;
+
+      tf2::doTransform(sensor_frame_position, map_frame_position, transform);
+
+      const Eigen::Vector2d measurement{map_frame_position.x(), map_frame_position.y()};
+
+      const Eigen::Matrix2d covariance = Eigen::Matrix2d::Identity() * 0.01;
```

Comment:
cov should come from the message

Reply:
Should is a strong word :)

If the covariance is in the message, should it reflect the covariance on the range/heading (as opposed to x/y)? And then should this node be doing math to convert it to x/y covariance for the filter?

Reply:
I think cov should be in x,y since that is what the message says but then convert it back out. I'm okay with just leaving it as this since that sounds more complex than it is probably worth though.

Reply:
The message doesn't use x/y, it uses range/heading.

Reply:
We'll revisit this when we get around to converting it into an exercise. It's not till week 6, so we've got time.